crates/mpt/src/node.rs

```diff
@@ -0,0 +1,191 @@
+//! This module contains the [TrieNode] type, which allows for decoding the RLP
+
+use alloc::{collections::VecDeque, vec::Vec};
+use alloy_primitives::Bytes;
+use alloy_rlp::{Buf, Decodable, EMPTY_LIST_CODE, EMPTY_STRING_CODE};
+use anyhow::{anyhow, Result};
+
+/// The length of the branch list when RLP encoded
+const BRANCH_LIST_LENGTH: usize = 17;
+
+/// The length of a leaf or extension node's RLP encoded list
+const LEAF_OR_EXTENSION_LIST_LENGTH: usize = 2;
+
+/// Prefix for even-nibbled extension node paths.
+const PREFIX_EXTENSION_EVEN: u8 = 0;
+
+/// Prefix for odd-nibbled extension node paths.
+const PREFIX_EXTENSION_ODD: u8 = 1;
+
+/// Prefix for even-nibbled leaf node paths.
+const PREFIX_LEAF_EVEN: u8 = 2;
+
+/// Prefix for odd-nibbled leaf node paths.
+const PREFIX_LEAF_ODD: u8 = 3;
+
+/// A [TrieNode] is a node within a standard Merkle Patricia Trie.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TrieNode {
+    /// A branch node refers to up to 16 child nodes with the encoding `rlp([ v0, ..., v15, value
+    /// ])`
+    Branch {
+        /// The 16 child nodes and value of the branch.
+        stack: VecDeque<NodeElement>,
+    },
+    /// A leaf node is a 2-item node with the encoding `rlp([encoded_path, value])`
+    Leaf {
+        /// The key of the leaf node
+        key: Bytes,
+        /// The value of the leaf node
+        value: Bytes,
+    },
+    /// An extension node is a 2-item pointer node with the encoding `rlp([encoded_path, key])`
+    Extension {
+        /// The path prefix of the extension
+        prefix: Bytes,
+        /// The pointer to the child node
+        node: Bytes,
+    },
```

Comment:
Can these types be restricted further than a generic Bytes type? ie something with a fixed size

Reply:
Both represent opaque RLP encoded values in both variants, as the `TrieNode` doesn't recursively deserialize. They're encoded as strings as well, so having them as `Bytes` makes them easy to store